src/CPUActivityAgent.cpp

```diff
@@ -136,46 +138,61 @@ namespace geopm
-                           std::to_string(in_policy.size()));
+        m_qm_max_rate = {};
+        const auto ALL_NAMES = m_platform_io.signal_names();
 
-        // Check for NAN to set default values for policy
-        if (std::isnan(in_policy[M_POLICY_CPU_FREQ_MAX])) {
-            in_policy[M_POLICY_CPU_FREQ_MAX] = m_freq_core_max;
+        // F efficient values
+        std::string fe_constconfig = "CONST_CONFIG::CPU_FREQUENCY_EFFICIENT_HIGH_INTENSITY";
+        if (ALL_NAMES.count(fe_constconfig) != 0) {
+            m_freq_core_efficient = m_platform_io.read_signal(fe_constconfig, GEOPM_DOMAIN_BOARD, 0);
         }
-
-        // Check for NAN to set default values for policy
-        if (std::isnan(in_policy[M_POLICY_CPU_FREQ_EFFICIENT])) {
-            in_policy[M_POLICY_CPU_FREQ_EFFICIENT] = m_freq_core_min;
+        else {
+            m_freq_core_efficient = m_freq_core_min;
         }
 
-        //////////////////////////
-        //UNCORE POLICY CHECKING//
-        //////////////////////////
-        if (std::isnan(in_policy[M_POLICY_UNCORE_FREQ_MAX])) {
-            in_policy[M_POLICY_UNCORE_FREQ_MAX] = m_freq_uncore_max;
+        fe_constconfig = "CONST_CONFIG::CPU_UNCORE_FREQUENCY_EFFICIENT_HIGH_INTENSITY";
+        if (ALL_NAMES.count(fe_constconfig) != 0) {
+            m_freq_uncore_efficient = m_platform_io.read_signal(fe_constconfig, GEOPM_DOMAIN_BOARD, 0);
         }
-        if (std::isnan(in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT])) {
-            in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] = m_freq_uncore_min;
+        else {
+            m_freq_uncore_efficient = m_freq_uncore_min;
         }
 
-        // When the policy is all NaNs, this check also verifies that
-        // the system was not left in a bad state with regard to the
-        // UNCORE_FREQUENCY_<MAX/MIN>_CONTROLs.
-        if (in_policy[M_POLICY_UNCORE_FREQ_EFFICIENT] > in_policy[M_POLICY_UNCORE_FREQ_MAX]) {
+        // Grab all (uncore frequency, max memory bandwidth) pairs
+        for (int entry_idx = 0; entry_idx < (int)ALL_NAMES.size(); ++entry_idx) {
+            const std::string KEY_NAME = "CONST_CONFIG::CPU_UNCORE_FREQUENCY_" +
+                                          std::to_string(entry_idx);
+            const std::string VAL_NAME = "CONST_CONFIG::CPU_UNCORE_MAX_MEMORY_BANDWIDTH_" +
+                                          std::to_string(entry_idx);
+            if (ALL_NAMES.find(KEY_NAME) != ALL_NAMES.end() &&
+                ALL_NAMES.find(VAL_NAME) != ALL_NAMES.end()) {
+                double uncore_freq = m_platform_io.read_signal(KEY_NAME, GEOPM_DOMAIN_BOARD, 0);
+                double max_mem_bw = m_platform_io.read_signal(VAL_NAME, GEOPM_DOMAIN_BOARD, 0);
+                if (!std::isnan(uncore_freq) && !std::isnan(max_mem_bw) &&
+                    uncore_freq != 0 && max_mem_bw != 0) {
+                    m_qm_max_rate[uncore_freq] = max_mem_bw;
+                }
```

Comment:
You know, I was thinking there must be a way to know how many pairs of uncore_freq-max_mem_bw are included in the configuration instead of having to sweep the entire domain of indices and seeing which names are valid. Would it be a terrible idea to just include that as one of the configuration parameters/signals in the ConstConfig file? Do any other more elegant ideas come to mind?

Reply:
Including a count signal of some sort would be a viable solution, but we have no way of enforcing accuracy (i.e. the count sig says 10 pairs, but there are 11) that aren't reliant upon doing the same sweep.  We could rely on the agent to tell us if there are less pairs than specified by the signal (i.e. missing signal throw from PlatformIO) but not more.  

Ultimately this is called once as part of init so I don't know if the search overhead is significant enough to be a concern.

Reply:
Yeah, that makes sense. I wasn't really concerned with the efficiency of the code, but rather its elegance and readability (because it took me a moment to realize why such a sweep was being done). However, even the idea I suggested would still require a similar sweep.